params/config.go

```diff
@@ -890,7 +890,13 @@ func (err *ConfigCompatError) Error() string {
 	if err.StoredBlock != nil {
 		return fmt.Sprintf("mismatching %s in database (have block %d, want block %d, rewindto block %d)", err.What, err.StoredBlock, err.NewBlock, err.RewindToBlock)
 	}
-	return fmt.Sprintf("mismatching %s in database (have timestamp %d, want timestamp %d, rewindto timestamp %d)", err.What, err.StoredTime, err.NewTime, err.RewindToTime)
+
+	if err.StoredTime == nil {
+		return fmt.Sprintf("mismatching %s in database (have timestamp nil, want timestamp %d, rewindto timestamp %d)", err.What, *err.NewTime, err.RewindToTime)
```

Comment:
This looks like a nil deref, via e.g. `new(ConfigCompatError).Error()`?

Reply:
Please add nil-checks every time  you deref one of the pointers

Reply:
if err.NewTime is nil, then will no ConfigCompatError Error
so err.NewTime is not nil

Reply:
```
[user@work go-ethereum]$ git diff 
diff --git a/params/config_test.go b/params/config_test.go
index bf8ce2fc5e..f6d0fd287c 100644
--- a/params/config_test.go
+++ b/params/config_test.go
@@ -137,3 +137,7 @@ func TestConfigRules(t *testing.T) {
                t.Errorf("expected %v to be shanghai", stamp)
        }
 }
+
+func TestFoo(t *testing.T) {
+       new(ConfigCompatError).Error()
+}
[user@work go-ethereum]$ go test ./params/ -run TestFoo
--- FAIL: TestFoo (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x527e22]

goroutine 5 [running]:
testing.tRunner.func1.2({0x5410e0, 0x674880})
        /usr/local/go/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1634 +0x377
panic({0x5410e0?, 0x674880?})
        /usr/local/go/src/runtime/panic.go:770 +0x132
```
What I mean is, your code is fragile. There is a publically exported type, `ConfigCompatError`, you are adding assumptions about how it is supposed to be used. And that's fine for complex things, but invoking the `Error()` is the most basic operation there is, you can't just make it crash

Reply:
> ```
> [user@work go-ethereum]$ git diff 
> diff --git a/params/config_test.go b/params/config_test.go
> index bf8ce2fc5e..f6d0fd287c 100644
> --- a/params/config_test.go
> +++ b/params/config_test.go
> @@ -137,3 +137,7 @@ func TestConfigRules(t *testing.T) {
>                 t.Errorf("expected %v to be shanghai", stamp)
>         }
>  }
> +
> +func TestFoo(t *testing.T) {
> +       new(ConfigCompatError).Error()
> +}
> [user@work go-ethereum]$ go test ./params/ -run TestFoo
> --- FAIL: TestFoo (0.00s)
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
>         panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x527e22]
> 
> goroutine 5 [running]:
> testing.tRunner.func1.2({0x5410e0, 0x674880})
>         /usr/local/go/src/testing/testing.go:1631 +0x24a
> testing.tRunner.func1()
>         /usr/local/go/src/testing/testing.go:1634 +0x377
> panic({0x5410e0?, 0x674880?})
>         /usr/local/go/src/runtime/panic.go:770 +0x132
> ```
> 
> What I mean is, your code is fragile. There is a publically exported type, `ConfigCompatError`, you are adding assumptions about how it is supposed to be used. And that's fine for complex things, but invoking the `Error()` is the most basic operation there is, you can't just make it crash

updated

Reply:
> ```
> [user@work go-ethereum]$ git diff 
> diff --git a/params/config_test.go b/params/config_test.go
> index bf8ce2fc5e..f6d0fd287c 100644
> --- a/params/config_test.go
> +++ b/params/config_test.go
> @@ -137,3 +137,7 @@ func TestConfigRules(t *testing.T) {
>                 t.Errorf("expected %v to be shanghai", stamp)
>         }
>  }
> +
> +func TestFoo(t *testing.T) {
> +       new(ConfigCompatError).Error()
> +}
> [user@work go-ethereum]$ go test ./params/ -run TestFoo
> --- FAIL: TestFoo (0.00s)
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
>         panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x527e22]
> 
> goroutine 5 [running]:
> testing.tRunner.func1.2({0x5410e0, 0x674880})
>         /usr/local/go/src/testing/testing.go:1631 +0x24a
> testing.tRunner.func1()
>         /usr/local/go/src/testing/testing.go:1634 +0x377
> panic({0x5410e0?, 0x674880?})
>         /usr/local/go/src/runtime/panic.go:770 +0x132
> ```
> 
> What I mean is, your code is fragile. There is a publically exported type, `ConfigCompatError`, you are adding assumptions about how it is supposed to be used. And that's fine for complex things, but invoking the `Error()` is the most basic operation there is, you can't just make it crash

updated